services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,202 @@
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+            throw new MessageException(error);
+        }
+
+        if (message instanceof ChunkedInfoMessage) {
+            processChunkedMessage((ChunkedInfoMessage) message);
+        } else if (pendingRequests.containsKey(message.getCorrelationId())) {
+            InfoMessage infoMessage = (InfoMessage) message;
+            CompletableFuture<InfoData> future = pendingRequests.remove(message.getCorrelationId());
+            future.complete(infoMessage.getData());
+        }
+    }
+
+    /**
+     * Performs searching and collecting all chunked messages into one chain if possible.
+     */
+    private synchronized void processChunkedMessage(ChunkedInfoMessage received) {
+        List<ChunkedInfoMessage> chain;
+        if (messagesChains.containsKey(received.getCorrelationId())) {
+            chain = messagesChains.remove(received.getCorrelationId());
+            chain.add(received);
+        } else if (pendingChunkedRequests.containsKey(received.getCorrelationId())) {
+            chain = Lists.newArrayList(received);
+        } else {
+            unlinkedChunks.put(received.getCorrelationId(), received);
+            return;
+        }
+
+        ChunkedInfoMessage tail = received;
+        // Check whether next messages for current chain is already received. If yes - add them into that chain.
+        while (tail.getNextRequestId() != null && unlinkedChunks.containsKey(tail.getNextRequestId())) {
+            chain.add(tail);
+            tail = unlinkedChunks.remove(tail.getNextRequestId());
+        }
+
+        if (tail.getNextRequestId() == null) {
+            // found last message in the chain
+            completeChain(chain);
+        } else {
+            messagesChains.put(tail.getNextRequestId(), chain);
+        }
+    }
+
+    /**
+     * Completes request.
+     */
+    private void completeChain(List<ChunkedInfoMessage> chain) {
+        List<InfoData> data = chain.stream()
+                .map(ChunkedInfoMessage::getData)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        ChunkedInfoMessage firstMessage = chain.get(0);
```

Comment:
If there is no any messages?

Reply:
There will be always at least one message, because it is a response from some component. And such message might have empty (null) payload (InfoData in our case)